query-algebrizer/src/clauses/not.rs

```diff
@@ -0,0 +1,450 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::btree_map::Entry;
+
+use mentat_core::{
+    Schema,
+};
+
+use mentat_query::{
+    NotJoin,
+    UnifyVars,
+    ContainsVariables,
+};
+
+use clauses::ConjoiningClauses;
+
+use errors::{
+    Result,
+};
+
+use types::{
+    ColumnConstraint,
+    ComputedTable,
+    EmptyBecause,
+};
+
+impl ConjoiningClauses {
+    pub fn apply_not_join(&mut self, schema: &Schema, not_join: NotJoin) -> Result<()> {
+        let projected = match not_join.unify_vars {
+            UnifyVars::Implicit => not_join.collect_mentioned_variables(),
+            UnifyVars::Explicit(vs) => vs.into_iter().collect(),
+        };
+        let mut template = self.use_as_template(&projected);
+        let mut empty_because: Option<EmptyBecause> = None;
+        
+        for clause in not_join.clauses.into_iter() {
+            template.apply_clause(&schema, clause)?;
+            // This is disgusting, but I can't think of anything else right now that works.
+            if template.is_known_empty() {
+                empty_because = template.empty_because;
+                template.empty_because = None;
+            }
+        }
+
+        if template.wheres.is_empty() && empty_because.is_some() {
+            return Ok(());
+        } else {
+            for (v, cols) in (&self.column_bindings).iter().filter(|&(key, _)| projected.contains(&key) ) {
+                match template.column_bindings.entry(v.clone()) {
+                    Entry::Vacant(e) => {
+                        e.insert(cols.clone());
+                    },
+                    Entry::Occupied(mut e) => {
+                        e.get_mut().append(&mut cols.clone());
+                    },
+                }
+            }
+        }
+
+        // We are only expanding column bindings here and not pruning extracted types as we are not projecting values.
```

Comment:
`all_datoms` is right — that just means we don't know if the value could be a string or not.

The missing part there is the type tag. `all_datoms02.v` might be `db.type/ref` = `25`, the entid, and `all_datoms01.v` might be a `db.type/long` = `25`, the number. Those are the same in SQLite, so we need to use the `value_type_tag` column to compare them — a ref will be `1` and a long will be `5`, so they won't unify.

It's possible that this will be entirely solved by #293; see the reference to that issue in `mod.rs`.